### PR TITLE
Wallet

### PR DESCRIPTION
**Pull request by lock9**
Hello,
This is our first concept for the wallet module. We are building the TransctionManager too, but we are still discussing some main concepts.

### CODE REVIEW

File: NeoSharp.TransactionManager/ITransactionManager.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            List<TransactionOutput> outputs,
+            double fee = 0);
+
+        Transaction BuildUnsignedTransaction(
+            IWalletAccount from,
+            List<TransactionAttribute> attributes,
+            List<TransactionOutput> outputs,
+            double fee = 0);
+
+
```
Comment (shargon) [inline NeoSharp.TransactionManager/ITransactionManager.cs]: Please tabule right the methods

File: NeoSharp.TransactionManager/ITransactionBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using NeoSharp.Core.Models;
+using NeoSharp.Core.Types;
+using Org.BouncyCastle.Math.EC;
+
+namespace NeoSharp.TransactionManager
+{
+	public interface ITransactionBuilder
```
Comment (shargon) [inline NeoSharp.TransactionManager/ITransactionBuilder.cs]: Tabule all right please

File: NeoSharp.TransactionManager/NEP-5/INEP5.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+using System.Text;
+
+namespace NeoSharp.TransactionManager.NEP_5
+{
+    interface INEP5
+    {
+		String TotalSupply();
+		String Name();
+		String Symbol();
+		String Decimals();
```
Comment (shargon) [inline NeoSharp.TransactionManager/NEP-5/INEP5.cs]: String => string
TotalSupply, Decimal and Balance return BigInteger
Transfer return bool 
Please ensure it on NEP proposals

File: NeoSharp.TransactionManager/NEP-5/INEP5TransactionManager.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace NeoSharp.TransactionManager
+{
+    public interface INEP5TransactionManager
+    {
+		Transaction BuildUnsignedNEP5Transaction(IWalletAccount from,
+									List<TransactionAttribute> attributes,
+									UInt160 output,
+									double fee = 0);
+
```
Comment (shargon) [inline NeoSharp.TransactionManager/NEP-5/INEP5TransactionManager.cs]: Tabule again :P

File: NeoSharp.TransactionManager/NeoSharp.TransactionManager.csproj [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (shargon) [inline NeoSharp.TransactionManager/NeoSharp.TransactionManager.csproj]: Why we need another project, maybe is better inside NeoSharp.Core

File: src/NeoSharp.Core/Models/TransactionAttribute.cs [hunk 6]
```diff
@@ -1,18 +1,20 @@
...
     {
-        [BinaryProperty(1)]
+        [BinaryProperty(0)]
         [JsonProperty("usage")]
         public TransactionAttributeUsage Usage;
 
-        [BinaryProperty(2)]
+        [BinaryProperty(1)]
         [JsonProperty("data")]
-        public string Data;
```
Comment (shargon) [inline src/NeoSharp.Core/Models/TransactionAttribute.cs:16]: Is hard to review with my previous changes :S

File: src/NeoSharp.Wallet/NEP6/NEP6Wallet.cs [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+			if (privateKey.Length != 32)
+			{
+				if (privateKey.Length == 96 || privateKey.Length == 104)
+				{
+					throw new NotImplementedException();
+				}
+				throw new ArgumentException();
+			}
+
+			X9ECParameters curve = SecNamedCurves.GetByName("secp256r1");
```
Comment (shargon) [inline src/NeoSharp.Wallet/NEP6/NEP6Wallet.cs]: Use ICrypto

File: src/NeoSharp.Wallet/NEP6/NEP6Wallet.cs [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+		}
+
+		/// <summary>
+		/// Creates the account with random private key.
+		/// </summary>
+		/// <returns>The account.</returns>
+		public IWalletAccount CreateAccount()
+		{
+			byte[] privateKey = new byte[32];
+			using (System.Security.Cryptography.RandomNumberGenerator rng =
```
Comment (shargon) [inline src/NeoSharp.Wallet/NEP6/NEP6Wallet.cs]: This should be moved to ICrypto

File: src/NeoSharp.Wallet/NeoSharp.Wallet.csproj [hunk 9]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
```
Comment (shargon) [inline src/NeoSharp.Wallet/NeoSharp.Wallet.csproj]: Maybe is better in NeoSharp.Core/Wallet ?

File: NeoSharp.TransactionManager/ITransactionManager.cs (hunk 1 above)
Comment (melanke) [reply NeoSharp.TransactionManager/ITransactionManager.cs]: Hi @shargon, are you following a specific code style? How this line should be indented?

File: src/NeoSharp.Application/Client/Prompt[Wallet].cs [hunk 10]
```diff
@@ -8,11 +10,20 @@ public partial class Prompt : IPrompt
...
         {
-            if (file.Exists)
+            IWallet newWallet;
+            try
+            {
+                newWallet = _wallet.CreateWallet(file);
+            }
+            catch(System.ArgumentException)
             {
                 _consoleWriter.WriteLine($"File '{file.FullName}' already exist, please provide a new one", ConsoleOutputStyle.Error);
```
Comment (shargon) [inline src/NeoSharp.Application/Client/Prompt[Wallet].cs]: I prefer the `if file.Exist` i don't know here if IWallet do this check

**Comment by gubanotorious** [comment]
Did this get migrated into the branch for working?  Please close this pull request if you are working in the branch...

**Comment by canesin** [comment]
@lock9 could you rebase and do a new clean PR ?

**Pull request closed without merging**